### PR TITLE
Change groupRandomly & groupAll to use OrderedSerialization

### PR DESCRIPTION
**Pull request by MansurAshraf**
@johnynek Not sure if this is the best approach but take a look

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala [hunk 1]
```diff
@@ -115,6 +118,7 @@ object TypedPipe extends Serializable {
  */
 trait TypedPipe[+T] extends Serializable {
 
+  private def primitiveOrderedBufferSupplier[T]: OrderedSerialization[T] = macro impl.OrderedSerializationProviderImpl[T]
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: what is `primitive` about this? Also, can we stop using the `OrderedBuffer` name which is the old name before `OrderedSerialization`. Why not `orderedSerialialization[T]`

**Comment by johnynek** [comment]
Didn't Gabe revert the merging of the fix for groupAll? Do you want to use the same trick to do groupAll in this PR?

**Comment by MansurAshraf** [comment]
Gabe will merge this and groupAll PR internally once its green. 






On Wed, Jun 3, 2015 at 5:42 PM -0700, "P. Oscar Boykin" <notifications@github.com> wrote:












Didn't Gabe revert the merging of the fix for groupAll? Do you want to use the same trick to do groupAll in this PR?

**Comment by Gabriel439** [comment]
I did revert the groupAll fix, but I'm going to apply it to the internal relase that I cut (and also open up a pull request to replace your old one)

**Comment by johnynek** [comment]
Okay. Well, this and the groupAll fix work differently. So, I don't think that is the ideal way to go. The groupAll fix avoided a dependency on the macros and by hand implemented the UnitOrderedSerialization again. If we are going to add the dependency as done here, why not use the same approach for the groupAll fix (then it becomes a 1 line change on top of this, rather than adding a new file, testing that new code and then using it).

**Comment by MansurAshraf** [comment]
Thats a fair point and if we are cool with dependency on serialization module I can do the same trick with groupAll, otherwise we will have to hand role OrderedSerialization for Int 




On Wed, Jun 3, 2015 at 5:55 PM -0700, "P. Oscar Boykin" <notifications@github.com> wrote:












Okay. Well, this and the groupAll fix work differently. So, I don't think that is the ideal way to go. The groupAll fix avoided a dependency on the macros and by hand implemented the UnitOrderedSerialization again. If we are going to add the dependency as done here, why not use the same approach for the groupAll fix (then it becomes a 1 line change on top of this, rather than adding a new file, testing that new code and then using it).

**Comment by johnynek** [comment]
I'd use the same trick (and as per #1308 we can probably merge that module into scalding-serialization).

**Comment by MansurAshraf** [comment]
@Gabriel439 I have added the fix for groupAll in this PR as well. Based on @johnynek suggestion, you will have to revert the other PR internally as well and just use this PR for both fixes

**Comment by Gabriel439** [comment]
Got it.  I will use your pull request exclusively for both changes.

**Comment by Gabriel439** [comment]
Is it possible that adding the serialization macros dependency caused the permgen failure?

**Comment by MansurAshraf** [comment]
I doubt it but I am not a macro expert

**Comment by johnynek** [comment]
Totally possible, but I guess it should be only used to create two new
classes, which should be small (but maybe we don't really know for sure).

On Thu, Jun 4, 2015 at 10:20 AM, Mansur Ashraf <notifications@github.com>
wrote:

> I doubt it but I am not a macro expert
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/pull/1307#issuecomment-109036443>.
>



-- 
Oscar Boykin :: @posco :: http://twitter.com/posco

**Comment by Gabriel439** [comment]
This time the error seems to be that one of the test suites ran too long (2 hours) -.-

I've run these tests for Mansur's branch locally successfully (i.e. everything up to and including commit 0cba5b0)

**Comment by MansurAshraf** [comment]
It seems like these new tests for [TypePipeDiff](https://github.com/twitter/scalding/blob/develop/scalding-core/src/test/scala/com/twitter/scalding/typed/TypedPipeDiffTest.scala) are pushing Travis over the edge.

stats on my MBP

Without TypedPipeDiffTest
```
real	1m51.315s
user	2m17.611s
sys	0m8.015s
```

With TypedPipeDiffTest
```
real	8m13.432s
user	6m52.950s
sys	0m11.701s
```

@isnotinvain is it possible that we can trim down some of these tests or are they all required??

File: scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala [hunk 2]
```diff
@@ -115,6 +118,7 @@ object TypedPipe extends Serializable {
  */
 trait TypedPipe[+T] extends Serializable {
 
+  private def orderedSerialialization[T]: OrderedSerialization[T] = macro impl.OrderedSerializationProviderImpl[T]
```
Comment (ianoc) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala]: Can we not get this from the package itself? not have a macros reference here

**Comment by ianoc** [comment]
LGTM, merge when green

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 3]
```diff
@@ -125,16 +125,6 @@ case class SketchJoined[K: Ordering, V, V2, R](left: Sketched[K, V],
       .withReducers(numReducers)
       .map{ case ((r, k), v) => (k, v) }
   }
-
-  private implicit def intKeyOrd: Ordering[(Int, K)] = {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: this looks like a merge fail. It is backing out the sketched fix.

**Comment by johnynek** [comment]
Other than the Sketched merge fail, this is fine with me.

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 4]
```diff
@@ -30,7 +30,7 @@ object Sketched {
 
   /**
    * This is based on the CMSHasherBigInt found in algebird (see docs for in depth explanation):
-   * https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala#L1086
+   * https://github.com/twitter/algebird/lob/develop/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala#L1086
```
Comment (isnotinvain) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: unintentional?

**Comment by isnotinvain** [comment]
@MansurAshraf maybe we can turn down the number of runs in the property tests for typed pipe diff? can that be done per-test?

**Comment by johnynek** [comment]
@isnotinvain yes, that can be done per test. We might only want 10 tries or something on an expensive test.

:+1:

**Comment by johnynek** [comment]
merge when green

**Comment by isnotinvain** [comment]
That's fine with me, however, 8 minutes is not 2 hours, so these tests are just one of many things slowing the build down right?

**Comment by MansurAshraf** [comment]
8 mins was on my mac which is much faster than the travis machine we get.

**Pull request merged**